api/tasks_test.go

```diff
@@ -0,0 +1,185 @@
+				So(errMsg, ShouldEqual, "Failed to find job that has the specified id")
+			})
+		})
+	})
+
+	Convey("Given an API that can create valid search reindex tasks and store their details in a Data Store", t, func() {
+		apiInstance := api.Setup(ctx, mux.NewRouter(), dataStorerMock, &apiMock.AuthHandlerMock{}, taskNames)
+		createTaskHandler := apiInstance.CreateTaskHandler(taskNames)
+
+		Convey("When the tasks endpoint is called to create and store a new reindex task", func() {
+			req := httptest.NewRequest("POST", fmt.Sprintf("http://localhost:25700/jobs/%s/tasks", validJobID1), bytes.NewBufferString(
+				fmt.Sprintf(createTaskPayloadFmt, emptyTaskName)))
+			req.Header.Set("Content-Type", "application/json")
+			req.Header.Set("Authorization", validServiceAuthToken)
+			resp := httptest.NewRecorder()
+
+			createTaskHandler.ServeHTTP(resp, req)
+
+			Convey("Then an empty search reindex job is returned with status code 400 because the task name is empty", func() {
+				So(resp.Code, ShouldEqual, http.StatusBadRequest)
+				errMsg := strings.TrimSpace(resp.Body.String())
+				So(errMsg, ShouldEqual, "invalid request body")
+			})
+		})
+	})
+
+	Convey("Given an API that can create valid search reindex tasks and store their details in a Data Store", t, func() {
+		apiInstance := api.Setup(ctx, mux.NewRouter(), dataStorerMock, &apiMock.AuthHandlerMock{}, taskNames)
+		createTaskHandler := apiInstance.CreateTaskHandler(taskNames)
+
+		Convey("When the tasks endpoint is called to create and store a new reindex task", func() {
+			req := httptest.NewRequest("POST", fmt.Sprintf("http://localhost:25700/jobs/%s/tasks", validJobID1), bytes.NewBufferString(
+				fmt.Sprintf(createTaskPayloadFmt, invalidTaskName)))
+			req.Header.Set("Content-Type", "application/json")
+			req.Header.Set("Authorization", validServiceAuthToken)
+			resp := httptest.NewRecorder()
+
+			createTaskHandler.ServeHTTP(resp, req)
+
+			Convey("Then an empty search reindex job is returned with status code 400 because the task name is invalid", func() {
+				So(resp.Code, ShouldEqual, http.StatusBadRequest)
+				errMsg := strings.TrimSpace(resp.Body.String())
+				So(errMsg, ShouldEqual, "invalid request body")
+			})
+		})
+	})
+}
+
+// Create Task Payload
+var createTaskPayloadFmt = `{
+	"task_name": "%s",
+	"number_of_documents": 5
+}`
+
+func ExpectedTask(jobID string, lastUpdated time.Time, numberOfDocuments int, taskName string) (models.Task, error) {
+	cfg, err := config.Get()
+	if err != nil {
+		return models.Task{}, fmt.Errorf("unable to retrieve service configuration: %w", err)
+	}
+	urlBuilder := url.NewBuilder("http://" + cfg.BindAddr)
+	job := urlBuilder.BuildJobURL(jobID)
+	self := urlBuilder.BuildJobTaskURL(jobID, taskName)
+	return models.Task{
+		JobID:       jobID,
+		LastUpdated: lastUpdated,
+		Links: &models.TaskLinks{
+			Self: self,
+			Job:  job,
+		},
+		NumberOfDocuments: numberOfDocuments,
+		TaskName:          taskName,
+	}, err
```

Comment:
Usually you would explicitly return nil here rather than err, helps with clarity and uniformity